### DISCUSSION TITLE
Long time to load csv files compared to Pandas

### DISCUSSION THREAD

**Discussion by eric-kaufmann** [atoti/atoti#781] [General]
Hi there,

For a project I was trying to load large csv files into atoti and noticed long loading times for the `session.read_csv()` function. So I did a little test comparing pandas with atoti. 

*(full script below)*

Using numpy, I created matrices of different sizes and saved them as csv files. The matrices are of size (128,n), so a matrix of size (128,1) consists of 128 doubles (8 bytes), so 1028 bytes = 1 KB.
``` python
kilobyte = np.ones((128,1)) * np.pi 
print(kilobyte.size * kilobyte.itemsize)

# 1024
```

As different matrix sizes I used
- 1 KB = 1 x 1 KB
- 512 KB = 512 x 1 KB
- 1 MB = 1024 x 1 KB
- 512 MB = 1024 x 512 x 1 KB
- 1 GB = 1024 x 1024 x 1 KB

```python
sizes = [1, 512, 1024, 1024*512, 1024*1024]
names = ['1KB', '512KB', '1MB', '512MB', '1GB']

# Generate datasets
for s,n in zip(sizes,names):
    filename = "data_sizes/datasize_"+str(n)+".csv"
    a = np.fromfunction(lambda i, j: i+j, (s, 128), dtype=float)  # dtype float = 64 bit
    np.savetxt(filename, a, delimiter=",")
```
*Disclaimer:* Because of the csv file structure, the csv file size is ~3 times larger than the matrix size. The reason is that one character is 1 byte in size, so to store a double (8 byte) you need 24 characters plus the comma (~25).  
Example: 0.00[...]00e00 -> ['0', '.', '0', ..., '0', '+', 'e', '0', '0']

In the next step, I first load all csv files into a variable using pandas followed by atoti. To get a better comparison and to avoid cache preferences, I loaded the csv files for each library in a separate for-loop and preheated the cache. 

```python
# Load csv files into pandas DataFrame
temp = pd.read_csv("data_sizes/datasize_1MB.csv", engine="pyarrow")  # heat up cache
for n in names:
    filename = "data_sizes/datasize_"+str(n)+".csv"
    
    start_pd = time.time_ns()
    df = pd.read_csv(filename, engine="pyarrow")
    end_pd = time.time_ns()

del temp
del df

# Load csv files into atoti Table
session = tt.Session()
temp = session.read_csv("data_sizes/datasize_1MB.csv", table_name='temp') # heat up cache
for n in names:
    filename = "data_sizes/datasize_"+str(n)+".csv"
    
    start_tt = time.time_ns()
    table = session.read_csv(filename, table_name=n)
    end_tt = time.time_ns()

del temp
del table
```

The results speak for themselves. The loading time for "large" csv files is quite long.

![loading_analysis](https://user-images.githubusercontent.com/56740185/234007006-f13619ff-64f7-4f73-b890-ddf3c0fd5ee1.png)

For comparison I used a [Macbook Air 2020 with M1 chip](https://en.wikipedia.org/wiki/MacBook_Air_(Apple_silicon)), atoti 0.7.4, pandas 1.5.2 in an anaconda enviroment.

Are there any possible optimizations for faster loading times (other than using different file types like parquet, etc.)?

Thanks for your help
Eric

**The full python script**
```python
import pandas as pd
import atoti as tt
import numpy as np
import time
import matplotlib.pyplot as plt
from os.path import exists
import seaborn as sns
from matplotlib.ticker import FormatStrFormatter

sizes = [1, 512, 1024, 1024*512, 1024*1024]
names = ['1KB', '512KB', '1MB', '512MB', '1GB']

# Generate datasets
for s,n in zip(sizes,names):
    print(n)
    filename = "data_sizes/datasize_"+str(n)+".csv"
    if exists(filename):
        continue
    a = np.fromfunction(lambda i, j: i+j, (s, 128), dtype=float)
    np.savetxt(filename, a, delimiter=",")

# Load csv files into pandas DataFrame
durations_pd = []
temp = pd.read_csv("data_sizes/datasize_1MB.csv", engine="pyarrow")  # heat up cache
print("Loading Time Pandas:")
for n in names:
    filename = "data_sizes/datasize_"+str(n)+".csv"
    
    start_pd = time.time_ns()
    df = pd.read_csv(filename, engine="pyarrow")
    end_pd = time.time_ns()
    
    duration = (end_pd - start_pd)/ (10 ** 9)
    durations_pd.append(duration)
    print(f"  - {n}: {duration}s")
del temp
del df

# Load csv files into atoti Table
session = tt.Session()
temp = session.read_csv("data_sizes/datasize_1MB.csv", table_name='temp') # heat up cache
durations_tt = []
print("Loading Time Atoti:")
for n in names:
    filename = "data_sizes/datasize_"+str(n)+".csv"
    
    start_tt = time.time_ns()
    table = session.read_csv(filename, table_name=n)
    end_tt = time.time_ns()
    
    duration = (end_tt - start_tt)/ (10 ** 9)
    durations_tt.append(duration)
    print(f"  - {n}: {duration}s")
del temp
del table

# Plot results
sns.set_theme(style="whitegrid", palette="pastel")

plt.plot(names, durations_pd,'-bo', label='pandas')
plt.plot(names, durations_tt,'-ro',label='atoti')
plt.title('Loading time of different sized matrices')
plt.xlabel('Matrix Size')
plt.ylabel('Loading Time')
plt.gca().yaxis.set_major_formatter(FormatStrFormatter('%ds'))
plt.legend(loc="upper left")

for n,d in zip(names,durations_pd):
    plt.annotate("{:.2f}".format(d),
                 (n,d), 
                 textcoords="offset points", 
                 xytext=(0,-15),
                 ha='center') 
    
for n,d in zip(names,durations_tt):
    plt.annotate("{:.2f}".format(d),
                 (n,d), 
                 textcoords="offset points", 
                 xytext=(0,10),
                 ha='center') 

    
plt.margins(0.1)
plt.grid(visible=True, axis='y')

fig1 = plt.gcf()
plt.show()
plt.draw()
fig1.savefig('loading_analysis.png', dpi=100)

```

**Comment by tibdex** [comment]
Hello Eric, thanks for this well written post and the full Python script. It helps a lot 👍.

We're actually in the process of adding performance-related options to `read/load_csv()`. We'll get back to you once these are available.

In the meantime, you can do:

```diff
+ df = pd.read_csv(filename, engine="pyarrow")
- table = session.read_csv(filename, table_name=n)
+ table = session.read_pandas(df, table_name=n)
```

On a MacBook Pro 2019 with a 2,3 GHz 8-Core Intel Core i9 it gives me:

![Screenshot 2023-04-25 at 11 50 26](https://user-images.githubusercontent.com/6574550/234333005-caae6ab6-3679-473f-ae9b-bdd98e55e73e.png)

so of course still slower than just `pd.read_csv()` but ~5x instead of ~10x.
